in-progress/8401-proof-timeliness/proof-timeliness.ipynb

```diff
@@ -0,0 +1,164 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from scipy.stats import hypergeom"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def calculate_probability(V, C):\n",
+    "    \"\"\"\n",
+    "    Calculate the probability of drawing more than 2/3 malicious validators\n",
+    "    in a committee of size C from a total validator set of size V.\n",
+    "\n",
+    "    :param V: Total number of validators\n",
+    "    :param C: Committee size\n",
+    "    :return: Probability\n",
+    "    \"\"\"\n",
+    "    malicious = V // 3  # Number of malicious validators (1/3 of total)\n",
+    "    threshold = int(2 * C // 3)  # Threshold for malicious majority\n",
+    "\n",
+    "    # Calculate the probability of having more than threshold malicious validators\n",
+    "    prob = 1 - hypergeom.cdf(threshold, V, malicious, C)\n",
+    "\n",
+    "    return prob\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def calculate_first_p_malicious(V, C, P):\n",
```

Comment:
For clarity down here I think it would have been neater using pmf a bit more instead of doing it manually. Only a readability thing really. 🫡 

```python
import numpy as np
from scipy.stats import hypergeom
# Problem statement 
# Draw a committee of 48 validators from a large set of 1000 validators, with 333 malicious validators
# We model this as a hypergeometric distribution (without replacement)
# Then from the smaller committee, compute the probability of drawing k 

N_VALIDATORS = 10000
N_MALICIOUS = N_VALIDATORS // 3
N_COMMITTEE = 48

k = 13

# Step 1: Probability distribution of number of malicious validators in the committee
# We'll calculate the probability of getting `mal_commitee` malicious validators in the committee
prob_malicious_in_committee = np.array(
    [hypergeom.pmf(mal_commitee, N_VALIDATORS, N_MALICIOUS, N_COMMITTEE) for mal_commitee in range(0, N_COMMITTEE+1)]
)

# Step 2: For each possible number of malicious validators in the committee, calculate the probability of drawing 
# malicious malicious validators
prob_drawing_all_malicious = np.zeros_like(prob_malicious_in_committee)
for mal_commitee in range(k, N_COMMITTEE+1):  # We need at least `k` malicious validators to have a non-zero probability
    prob_drawing_all_malicious[mal_commitee] = hypergeom.pmf(k, N_COMMITTEE, mal_commitee, k)

# Total probability: Sum the probabilities of each scenario
total_prob = np.sum(prob_malicious_in_committee * prob_drawing_all_malicious)

print(f"The probability of drawing {k} malicious validators in the first {k} draws from the committee is: {total_prob}")
```